addons/sale_stock/tests/test_sale_stock_lead_time.py

```diff
@@ -98,3 +98,163 @@ def test_01_product_route_level_delays(self):
+            'partner_shipping_id': self.partner_1.id,
+            'pricelist_id': self.env.ref('product.list0').id,
+            'picking_policy': 'direct',
+            'warehouse_id': self.warehouse_1.id,
+            'order_line': [(0, 0, {'name': self.product_1.name,
+                                   'product_id': self.product_1.id,
+                                   'product_uom_qty': 5,
+                                   'product_uom': self.uom_unit.id,
+                                   'customer_lead': self.product_1.sale_delay})]})
+
+        # Confirm our standard sale order
+        order.action_confirm()
+
+        # Check the picking crated or not
+        self.assertTrue(order.picking_ids, "Pickings should be created.")
+
+        # Check schedule date of ship type picking
+        out = order.picking_ids.filtered(lambda r: r.picking_type_id == self.warehouse_1.out_type_id)
+        out_min_date = fields.Datetime.from_string(out.scheduled_date)
+        out_date = fields.Datetime.from_string(order.date_order) + timedelta(days=self.product_1.sale_delay) - timedelta(days=out.move_lines[0].rule_id.delay)
+        self.assertTrue(abs(out_min_date - out_date) <= timedelta(seconds=1), 'Schedule date of ship type picking should be equal to: order date + Customer Lead Time - pull rule delay.')
+
+        # Check schedule date of pack type picking
+        pack = order.picking_ids.filtered(lambda r: r.picking_type_id == self.warehouse_1.pack_type_id)
+        pack_min_date = fields.Datetime.from_string(pack.scheduled_date)
+        pack_date = out_date - timedelta(days=pack.move_lines[0].rule_id.delay)
+        self.assertTrue(abs(pack_min_date - pack_date) <= timedelta(seconds=1), 'Schedule date of pack type picking should be equal to: Schedule date of ship type picking - pull rule delay.')
+
+        # Check schedule date of pick type picking
+        pick = order.picking_ids.filtered(lambda r: r.picking_type_id == self.warehouse_1.pick_type_id)
+        pick_min_date = fields.Datetime.from_string(pick.scheduled_date)
+        pick_date = pack_date - timedelta(days=pick.move_lines[0].rule_id.delay)
+        self.assertTrue(abs(pick_min_date - pick_date) <= timedelta(seconds=1), 'Schedule date of pick type picking should be equal to: Schedule date of pack type picking - pull rule delay.')
+
+        # Now change the schedule date of pick
+        # Note : pack and out has change scheduled_date automatically based on delay set on pick
+        pick.write({'scheduled_date': pick_min_date + timedelta(days=5)})
+
+        # Now check scheduled_date of pack and out are changed or not based on propagate is true on rules?
+        self.assertEquals(pack.scheduled_date, (pack_min_date + timedelta(days=5)),
+            'Schedule date of pack should be changed based on delay.')
+        self.assertEquals(out.scheduled_date, (out_min_date + timedelta(days=5)),
+            'Schedule date of out should be changed based on delay.')
```

Comment:
complete this test, simulate _action_done on the pack by writing on on it date + timedeldtay 10 and state done, check the out_date is incremented by 10 days

Reply:
actually no, just add the tests i said in stock

add a test showing that po lines aren't merged if they have different expected dates